File: dmlab2d/game_scripts/common/properties.lua
--[[ Copyright (C) 2018-2019 The DMLab2D Authors.

Licensed under the Apache License, Version 2.0 (the 'License');
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an 'AS IS' BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
]]

local strings = require 'common.strings'
local system_properties = require 'system.properties'
local RESULT = system_properties

--[[ Returns the head and tail of keyList.

Examples:

> print(headTailSplit('aaa.bbb.ccc'))
'aaa'    'bbb.ccc'

> print(headTailSplit('aaa'))
'aaa'    nil

> print(headTailSplit('10.12'))
10    '12'

Arguments:

*   keyList - String containing dot (.) separated list of keys.

Returns:

*   Arg1 - First element of keyList (converted to a number if possible).
*   Arg2 - Tail of keyList still dot(.) separated. Returns nil if empty.
]]
local function headTailSplit(keyList)
  local dotIdx = keyList:find('[.]')
  if dotIdx then
    local key = keyList:sub(1, dotIdx - 1)
    key = tonumber(key) or key
    return key, keyList:sub(dotIdx + 1)
  end
  return strings.convertFrom(keyList), nil
end

--[[ Returns value of properties in keyList context.

See implmentation for details.

Arguments:

*   properties - table|function|string
*   keyList - String containing dot (.) separated list of keys.

Returns:

*   String if lookup is successful, RESULT.PERMISSION_DENIED if not not
    readable, otherwise RESULT.NOT_FOUND if not findable. If nil
    is returned it is equivalent to returning RESULT.NOT_FOUND.

Examples:

```lua
-- Look up in tables.
properties = {abc = {a = 1, b = 2, c = 3}}
assert(readProperty(properties, 'abc.b') == '2')
assert(readProperty(properties, 'abc.d') == RESULT.NOT_FOUND)

-- Look up in arrays.
properties = {abc = {'a', 'b', 'c'}}
assert(readProperty(properties, 'abc.2') == 'b')
assert(readProperty(properties, 'abc.0') == RESULT.NOT_FOUND)

-- Look up in functions.
properties = {abc = function(listKeys) return listKeys end}
assert(readProperty(properties, 'abc.hello_world'), 'hello_world')
```
]]
local function readProperty(properties, keyList)
  if properties == nil then
    return RESULT.NOT_FOUND
  end
  local propType = type(properties)
  if keyList == nil or keyList == '' then
    if propType == 'table' then
      return RESULT.PERMISSION_DENIED
    elseif propType == 'function' then
      return properties(keyList)
    else
      return tostring(properties)
    end
  end
  if propType == 'table' then
    local head, tail = headTailSplit(keyList)
    return readProperty(properties[head], tail)
  elseif propType == 'function' then
    return properties(keyList)
  else
    return RESULT.PERMISSION_DENIED
  end
end

--[[ Write value to writable in keyList context.

See implmentation for details.

Arguments:

*   properties - table|function
*   keyList - String containing dot (.) separated list of keys.

Returns:

*   RESULT.SUCCESS if write is successful, RESULT.INVALID_ARGUMENT if value is
    the wrong type, RESULT.PERMISSION_DENIED if the object is not writable or
    RESULT.NOT_FOUND if the property is not found. If nil is returned it is
    equivalent to returning RESULT.NOT_FOUND.

Examples:
```lua
-- Write number.
params = {abc = {a = 1, b = 2, c = 3}}
assert(writeProperty(params, 'abc.b', 22), RESULT.SUCCESS)
assert(writeProperty(params, 'abc.b', 'blah'), RESULT.INVALID_ARGUMENT)
assert(writeProperty(params, 'abc.d', 'blah'), RESULT.NOT_FOUND)

-- Write to array.
params = {abc = {'a', 'b', 'c'}}
assert(writeProperty(params, 'abc.2', 'bb'), RESULT.SUCCESS)
assert(writeProperty(params, 'abc.0'), RESULT.NOT_FOUND)
assert(writeProperty(params, 'abc', ''), RESULT.PERMISSION_DENIED)

-- Write to function
local params2 = {}
params2.abc = {
    add = function(key, value)
      params2.abc[key] = value
      return RESULT.SUCCESS
    end
}
assert(writeProperty(params2, 'abc.add.newKey', 'newValue'), RESULT.SUCCESS)
assert(params2.abc.newKey == 'newValue')
```
]]
local function writeProperty(properties, keyList, value)
  if properties == nil then
    return RESULT.NOT_FOUND
  end
  local t = type(properties)
  if t == 'table' then
    if keyList == nil then
      return RESULT.PERMISSION_DENIED
    end
    local head, tail = headTailSplit(keyList)
    if tail == nil then
      local val = properties[head]
      if val == nil then
        return RESULT.NOT_FOUND
      end
      local tval = type(val)
      if tval == 'function' then
        return pcall(val, value) and RESULT.SUCCESS or RESULT.INVALID_ARGUMENT
      elseif tval == 'string' then
        properties[head] = value
        return RESULT.SUCCESS
      elseif tval == 'number' then
        value = tonumber(value)
        if value ~= nil then
          properties[head] = value
          return RESULT.SUCCESS
        else
          return RESULT.INVALID_ARGUMENT
        end
      elseif tval == 'boolean' then
        if value == 'true' then
          properties[head] = true
          return RESULT.SUCCESS
        elseif value == 'false' then
          properties[head] = false
          return RESULT.SUCCESS
        else
          return RESULT.INVALID_ARGUMENT
        end
      else
        return RESULT.PERMISSION_DENIED
      end
    end
    return writeProperty(properties[head], tail, value)
  elseif t == 'function' then
    return properties(keyList, value) or RESULT.INVALID_ARGUMENT
  else
    return RESULT.PERMISSION_DENIED
  end
end

--[[ Add property at location keyList with value value.

If keyList is split into key1, key2 ... keyN then
properties[key1][key2]...[keyN] is assinged to value.

If properties[key1][key2]...[keyN] already exists or any part is not a table
then the operation is not successful and the value is not assigned. Use
writeProperty instead. Returns whether the operation was successful.
]]
local function addProperty(properties, keyList, value)
  if type(properties) ~= 'table' then
    return false
  end
  local head, tail = headTailSplit(keyList)
  local property = properties[head]
  if tail == nil then
    if property == nil then
      properties[head] = value
      return true
    else
      return false
    end
  else
    if property == nil then
      property = {}
      properties[head] = property
    end
    return addProperty(property, tail, value)
  end
end

local function removeProperty(properties, keyList)
  if type(properties) ~= 'table' then
    return false
  end
  local head, tail = headTailSplit(keyList)
  local property = properties[head]
  if tail == nil then
    if property ~= nil then
      properties[head] = nil
      return true
    else
      return false
    end
  else
    if property == nil then
      return false
    end
    return removeProperty(property, tail)
  end
end

local _readWrite = {}
local _readOnly = {}
local _writeOnly = {}

local function addReadWrite(keyList, property)
  addProperty(_readWrite, keyList, property)
end

local function addReadOnly(keyList, property)
  addProperty(_readOnly, keyList, property)
end

local function addWriteOnly(keyList, property)
  addProperty(_writeOnly, keyList, property)
end

local function removeReadWrite(keyList)
  removeProperty(_readWrite, keyList)
end

local function removeReadOnly(keyList)
  removeProperty(_readOnly, keyList)
end

local function removeWriteOnly(keyList)
  removeProperty(_writeOnly, keyList)
end

local function apiWriteProperty(_, keyList, value)
  local resRW = writeProperty(_readWrite, keyList, value)
  if resRW == RESULT.SUCCESS then
    return resRW
  end
  local resW0 = writeProperty(_writeOnly, keyList, value)
  if resW0 == RESULT.SUCCESS then
    return resW0
  end

  -- If result is not a success return the most specific error message.
  return math.max(resRW or RESULT.NOT_FOUND,
                  resW0 or RESULT.NOT_FOUND)
end

local function apiReadProperty(_, keyList)
  local resRW = readProperty(_readWrite, keyList)
  if resRW ~= nil and type(resRW) ~= 'number' then
    return tostring(resRW)
  end
  local resRO = readProperty(_readOnly, keyList)
  if resRO ~= nil and type(resRO) ~= 'number' then
    return tostring(resRO)
  end

  -- If result is not a success return the most specific error message.
  return math.max(resRW or RESULT.NOT_FOUND,
                  resRO or RESULT.NOT_FOUND)
end

local function apiListProperty(_, keyList, callback)
  local localReadWrite = _readWrite
  local localReadOnly = _readOnly
  local localWriteOnly = _writeOnly
  local prefix = ''
  if keyList == '' then
    keyList = nil
  end
  -- Avoid calling any functions for listing.
  while keyList do
    local head, tail = headTailSplit(keyList)
    prefix = prefix .. head .. '.'
    keyList = tail
    localReadWrite = localReadWrite and
                     type(localReadWrite) == 'table' and
                     localReadWrite[head]
    localReadOnly = localReadOnly and
                     type(localReadOnly) == 'table' and
                     localReadOnly[head]
    localWriteOnly = localWriteOnly and
                     type(localWriteOnly) == 'table' and
                     localWriteOnly[head]
  end
  if localReadWrite == nil and
     localReadOnly == nil and
     localWriteOnly == nil then
     return RESULT.NOT_FOUND
  end
  local status = RESULT.NOT_FOUND
  if localReadWrite and type(localReadWrite) == 'table' then
    status = RESULT.SUCCESS
    for k, v in pairs(localReadWrite) do
      callback(prefix .. k, type(v) == 'table' and 'l' or 'rw')
    end
  end

  if localReadOnly and type(localReadOnly) == 'table' then
    status = RESULT.SUCCESS
    for k, v in pairs(localReadOnly) do
      callback(prefix .. k, type(v) == 'table' and 'l' or 'r')
    end
  end

  if localWriteOnly and type(localWriteOnly) == 'table' then
    status = RESULT.SUCCESS
    for k, v in pairs(localWriteOnly) do
      callback(prefix .. k, type(v) == 'table' and 'l' or 'w')
    end
  end
  return status
end


--[[ Adds writeProperty, readProperty and listProperty to an API.

Arguments:

*   `api` An environment API table without `writeProperty`, `readProperty` and
    `listProperty`. Exposes properties added via calls to `addReadWrite`,
    `addReadOnly` and `addWriteOnly`.
]]
local function decorate(api)
  assert(api.writeProperty == nil, 'Already has writeProperty')
  assert(api.readProperty == nil, 'Already has readProperty')
  assert(api.listProperty == nil, 'Already has listProperty')
  api.writeProperty = apiWriteProperty
  api.readProperty = apiReadProperty
  api.listProperty = apiListProperty
  return api
end

return {
    RESULT = RESULT,
    decorate = decorate,
    addReadWrite = addReadWrite,
    addReadOnly = addReadOnly,
    addWriteOnly = addWriteOnly,
    removeReadWrite = removeReadWrite,
    removeReadOnly = removeReadOnly,
    removeWriteOnly = removeWriteOnly,
}

File: dmlab2d/lib/env_lua_api/properties.cc
// Copyright (C) 2018-2019 The DMLab2D Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

#include "dmlab2d/lib/env_lua_api/properties.h"

#include <string>

#include "absl/log/check.h"
#include "absl/log/log.h"
#include "absl/strings/match.h"
#include "absl/strings/string_view.h"
#include "dmlab2d/lib/lua/bind.h"
#include "dmlab2d/lib/lua/call.h"
#include "dmlab2d/lib/lua/lua.h"
#include "dmlab2d/lib/lua/n_results_or.h"
#include "dmlab2d/lib/lua/push.h"
#include "dmlab2d/lib/lua/read.h"
#include "dmlab2d/lib/lua/stack_resetter.h"
#include "dmlab2d/lib/lua/table_ref.h"
#include "third_party/rl_api/env_c_api.h"

namespace deepmind::lab2d {
namespace {

EnvCApi_PropertyResult ProcessResult(lua_State* L, lua::NResultsOr result,
                                     absl::string_view call) {
  if (result.n_results() == 0) {
    LOG_IF(ERROR, !result.ok()) << "[" << call << "] - " << result.error();
    return EnvCApi_PropertyResult_PermissionDenied;
  }
  int prop_result;
  if (IsFound(lua::Read(L, 1, &prop_result))) {
    switch (prop_result) {
      case EnvCApi_PropertyResult_Success:
        return EnvCApi_PropertyResult_Success;
      case EnvCApi_PropertyResult_NotFound:
        return EnvCApi_PropertyResult_NotFound;
      case EnvCApi_PropertyResult_PermissionDenied:
        return EnvCApi_PropertyResult_PermissionDenied;
      case EnvCApi_PropertyResult_InvalidArgument:
        return EnvCApi_PropertyResult_InvalidArgument;
    }
  }

  LOG(ERROR) << "[" << call
             << "] - "
                "- Invalid return type from write! "
                "Must return integer in range [0, 3] - "
                "0 (properties.SUCCESS), "
                "1 (properties.NOT_FOUND), "
                "2 (properties.PERMISSION_DENIED), "
                "3 (properties.INVALID_ARGUMENT)";
  return EnvCApi_PropertyResult_PermissionDenied;
}

struct PropertyListCallbackData {
  void* userdata;
  void (*call)(void* userdata, const char* key,
               EnvCApi_PropertyAttributes flags);
};

lua::NResultsOr PropertyListCallBackFunction(lua_State* L) {
  auto* callback_data = static_cast<PropertyListCallbackData*>(
      lua_touserdata(L, lua_upvalueindex(1)));
  std::string key;
  absl::string_view mode;

  if (!IsFound(lua::Read(L, 1, &key))) {
    return "[propertyList.callback] - Missing Arg 1 - Key";
  }
  if (!IsFound(lua::Read(L, 2, &mode))) {
    return "[propertyList.callback] -  Missing Arg 2 - Mode";
  }

  for (auto c : mode) {
    switch (c) {
      case 'w':
      case 'r':
      case 'l':
        break;
      default:
        return "Type mismatch mode must in the format [r][w][l] of: "
               " 'r' - read-only,"
               " 'w' - write-only,"
               " 'l' - listable";
    }
  }

  int flags = 0;
  if (absl::StrContains(mode, 'w')) {
    flags |= EnvCApi_PropertyAttributes_Writable;
  }
  if (absl::StrContains(mode, 'r')) {
    flags |= EnvCApi_PropertyAttributes_Readable;
  }
  if (absl::StrContains(mode, 'l')) {
    flags |= EnvCApi_PropertyAttributes_Listable;
  }
  if (flags != 0) {
    callback_data->call(callback_data->userdata, key.c_str(),
                        static_cast<EnvCApi_PropertyAttributes>(flags));
  }
  return 0;
}

}  // namespace

lua::NResultsOr Properties::Module(lua_State* L) {
  auto module = lua::TableRef::Create(L);
  module.Insert("SUCCESS", static_cast<int>(EnvCApi_PropertyResult_Success));
  module.Insert("NOT_FOUND", static_cast<int>(EnvCApi_PropertyResult_NotFound));
  module.Insert("PERMISSION_DENIED",
                static_cast<int>(EnvCApi_PropertyResult_PermissionDenied));
  module.Insert("INVALID_ARGUMENT",
                static_cast<int>(EnvCApi_PropertyResult_InvalidArgument));
  lua::Push(L, module);
  return 1;
}

EnvCApi_PropertyResult Properties::WriteProperty(const char* key,
                                                 const char* value) {
  lua_State* L = script_table_ref_.LuaState();
  lua::StackResetter stack_resetter(L);
  script_table_ref_.PushMemberFunction("writeProperty");
  if (lua_isnil(L, -2)) {
    return EnvCApi_PropertyResult_NotFound;
  }
  lua::Push(L, key);
  lua::Push(L, value);
  return ProcessResult(L, lua::Call(L, 3), "writeProperty");
}

EnvCApi_PropertyResult Properties::ReadProperty(const char* key,
                                                const char** value) {
  lua_State* L = script_table_ref_.LuaState();
  lua::StackResetter stack_resetter(L);
  script_table_ref_.PushMemberFunction("readProperty");
  if (lua_isnil(L, -2)) {
    return EnvCApi_PropertyResult_NotFound;
  }
  lua::Push(L, key);
  auto result = lua::Call(L, 2);
  if (result.n_results() == 1) {
    if (IsFound(lua::Read(L, 1, &property_storage_))) {
      *value = property_storage_.c_str();
      return EnvCApi_PropertyResult_Success;
    }
  }

  *value = "";
  return ProcessResult(L, result, "readProperty");
}

EnvCApi_PropertyResult Properties::ListProperty(
    void* userdata, const char* list_key,
    void (*prop_callback)(void* userdata, const char* key,
                          EnvCApi_PropertyAttributes flags)) {
  lua_State* L = script_table_ref_.LuaState();
  lua::StackResetter stack_resetter(L);
  script_table_ref_.PushMemberFunction("listProperty");
  if (lua_isnil(L, -2)) {
    return EnvCApi_PropertyResult_NotFound;
  }
  lua::Push(L, list_key);
  PropertyListCallbackData callback_data{userdata, prop_callback};
  lua_pushlightuserdata(L, &callback_data);
  lua_pushcclosure(L, lua::Bind<PropertyListCallBackFunction>, 1);
  return ProcessResult(L, lua::Call(L, 3), "listProperty");
}

}  // namespace deepmind::lab2d
